### DISCUSSION TITLE
jupyter notebook can run, but nothing happened on screen

### DISCUSSION THREAD

**Discussion by cwrueciv** [coder/code-server#4837] [Q&A]
Hi, I just setup the code-server and add the Python extension. But when I open a jupyter notebook file, nothing shows on the screen. However, I do can select an interpreter. And I also can click the 'Run All' button. Since the notebook contains plt.savefig(...), I can find new images generated in the remote server folder after I hit the 'Run All'.  I can also hit 'restart kernel', 'clear outputs of all cells', 'Restart', etc. It seems everything is functional except nothing shows on screen.

Also, 'Papyrus PDF Preview' does not work as well. Not sure if it is related.

Here is a screenshot. Similar to this [issue](https://github.com/microsoft/vscode-jupyter/issues/1062)?

<img width="1696" alt="error1" src="https://user-images.githubusercontent.com/98050432/152903776-2ac4467b-6314-4d49-86ee-271d7274e7a7.png">

Any suggestions?

**Comment by cwrueciv** [comment]
It looks like it is because I’m using http rather than https. Hope there is a method to work around

**Reply by lg-nineoneninethreenine** [reply]
I think you could get a Let's Encrypt cert on your instance. There's documentation at Coder.com, so you can see that!

**Reply by jsjoeio** [reply]
We have some docs here: [test/guide#https-and-self-signed-certificates](https://coder.com/docs/code-server/latest/guide#https-and-self-signed-certificates)

**Reply by cwrueciv** [reply]
Thanks for your suggestion! The problem is I am working inside my university's local area network. So I probably cannot buy a domain name and expose it to the internet? Is there a method that I can use https in the local area network? If I am at home I can just use the VPN to connect to my server.

**Reply by jsjoeio** [reply]
Hmm...maybe you could use [`ngrok`](https://ngrok.com/) or [`localtunnel`](https://www.npmjs.com/package/localtunnel) to expose it? Would that work?

**Reply by cwrueciv** [reply]
Hi, I think I can expose it by a program named [nps](https://github.com/ehang-io/nps). I am not an expert about expose, but I think it probably does similar work to ngrok or localtunnel?

So what should I do after expose? I think I can expose the port 8080 of my local computer to a Google server. . For example, if I visit xx.xx.xx.xx:1234(my google server, http) will be equivalent to visiting my local computer xx.xx.xx.xx:8080 (the code-server). So can I use the NGINX or Caddy? If it works, should I add my local computer's IP or my Google server's IP to the domain (for example, Google domain)? Most importantly, what should I do with the port number? Because only xx.xx.xx.xx:1234 will point to my code-server. 

If exposing to the server is too complex, working inside the VPN is also good for me (if there is any method). 

Thanks so much!

**Reply by cwrueciv** [reply]
A quick update. I used the google domain and the nps to forward the code-server, which is 127.0.0.1:8080. Currently I can access code-server, but it shows the error below. 
<img width="962" alt="屏幕截图 2022-02-09 141905" src="https://user-images.githubusercontent.com/98050432/153274235-5f22b243-03f6-4aeb-a8eb-9122014e12a2.png">

But when I use ssh -N -L to visit the 127.0.0.1:8080, everything works.

**Reply by jsjoeio** [reply]
Hmm...unfortunately I'm not an expert here so I can't help much. But we do have some docs you can reference: https://coder.com/docs/code-server/latest/guide#expose-code-server

**Reply by cwrueciv** [reply]
Thank you! I think there is no easy way to work around. I am going to leave a few notes here in case anyone interested. 

1) The jupyter notebook seems cannot be opened via http.

2) Using intranet penetration proxy server probably will work, but you need **WebSocket**. The tool I used does not support websocket. 

3) The only way works for me is using the port forwarding, i.e. ssh -L -N.........

4) Another work around is directly using a jupyter notebook [server](https://jupyter-notebook.readthedocs.io/en/stable/public_server.html). So you don't need any https. It can be visited via TCP

@jsjoeio I am just wondering why we cannot use jupyter notebook funtion in http?

**Reply by jsjoeio** [reply]
> @jsjoeio I am just wondering why we cannot use jupyter notebook funtion in http?

That's a limitation of VS Code, or rather the browser. I believe VS Code uses service workers to render webviews (like Jupyter Notebook). Browser do not allow service workers in insecure environments. Therefore, you must use HTTPS.

You may be able to override something at the browser level though (i.e. [here is an option in Brave](brave://flags/%23allow-insecure-localhost#unsafely-treat-insecure-origin-as-secure)). I haven't tested though, just a thought.
